src/libcore/va_list.rs

```diff
@@ -0,0 +1,88 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Implementation of a `va_list`.
+
+#[cfg(target_arch = "aarch64")]
+#[repr(C)]
+#[derive(Debug)]
+/// AArch64 ABI implementation of a `va_list`. See the
+/// [Aarch64 Procedure Call Standard] for more details.
+///
+/// [AArch64 Procedure Call Standard]:
+/// http://infocenter.arm.com/help/topic/com.arm.doc.ihi0055b/IHI0055B_aapcs64.pdf
+pub struct VaListImpl {
+    stack: *mut (),
+    gr_top: *mut (),
+    vr_top: *mut (),
+    gr_offs: i32,
+    vr_offs: i32,
+}
+
+#[cfg(any(target_arch = "arm", target_arch = "mips", target_arch = "powerpc64",
+          target_arch = "x86", all(target_arch = "x86_64", windows)))]
+/// Basic byte pointer variant of a `va_list`.
+pub type VaListImpl = *const i8;
+
+#[cfg(target_arch = "powerpc")]
+#[repr(C)]
+#[derive(Debug)]
+/// PowerPC ABI implementation of a `va_list`.
+pub struct VaListImpl {
+    gpr: u8,
+    fpr: u8,
+    reserved: u16,
+    overflow_arg_area: *mut (),
+    reg_save_area: *mut (),
+}
+
+#[cfg(all(target_arch = "x86_64", not(windows)))]
+#[repr(C)]
+#[derive(Debug)]
+/// x86_64 ABI implementation of a `va_list`.
+pub struct VaListImpl {
+    gp_offset: i32,
+    fp_offset: i32,
+    overflow_arg_area: *mut (),
+    reg_save_area: *mut (),
+}
+
+/// A wrapper for a `va_list`
+#[cfg(not(stage0))]
+#[lang = "va_list"]
+#[unstable(feature = "c_variadic",
+           reason = "the `c_variadic` feature has not been properly tested on \
+                     all supported platforms",
+           issue = "27745")]
+#[derive(Debug)]
+pub struct VaList<'a>(&'a mut VaListImpl);
+
+#[cfg(not(stage0))]
+#[unstable(feature = "c_variadic",
+           reason = "the `c_variadic` feature has not been properly tested on \
+                     all supported platforms",
+           issue = "27745")]
+impl<'a> VaList<'a> {
+    /// Advance to the next arg.
+    pub unsafe fn arg<T: Copy>(&mut self) -> T {
+        ::intrinsics::va_arg(self.0)
+    }
+
+    /// Copy the `va_list` at the current location.
+    pub unsafe fn copy<'ret, F, T>(&self, f: F) -> T
+            where F: FnOnce(&mut VaList) -> T, T: 'ret {
+        let mut ap_tag = ::mem::uninitialized::<VaListImpl>();
```

Comment:
Can you avoid `tag` in names? Also, what is `ap`?

Reply:
Yeah, `ap` is a standard arg name for `va_list`'s e.g. https://linux.die.net/man/3/va_start